src/rgw/rgw_rados.cc

```diff
@@ -11967,6 +11961,73 @@ int RGWRados::get_bucket_info(RGWObjectCtx& obj_ctx,
   return 0;
 }
 
+int RGWRados::get_bucket_info(RGWObjectCtx& obj_ctx,
+                              const string& tenant, const string& bucket_name,
+                              RGWBucketInfo& info,
+                              real_time *pmtime, map<string, bufferlist> *pattrs)
+{
+  bucket_info_entry e;
+  string bucket_entry;
+  rgw_make_bucket_entry_name(tenant, bucket_name, bucket_entry);
+
+  if (binfo_cache->find(bucket_entry, &e)) {
+    info = e.info;
+    if (pattrs)
+      *pattrs = e.attrs;
+    if (pmtime)
+      *pmtime = e.mtime;
+    return 0;
+  }
+
+  return bucket_info_cache_miss(obj_ctx, tenant, bucket_name, info, pmtime,
+                                pattrs);
+}
+
+bool RGWRados::try_refresh_bucket_info(RGWBucketInfo& info,
+                                       ceph::real_time *pmtime,
+                                       map<string, bufferlist> *pattrs)
+{
+  RGWObjectCtx obj_ctx(this);
+  bucket_info_entry e;
+  string bucket_entry;
+  rgw_make_bucket_entry_name(info.bucket.tenant, info.bucket.name, bucket_entry);
+
+  auto checkv = info.objv_tracker.read_version;
+  if (binfo_cache->find(bucket_entry, &e) &&
+      !e.info.objv_tracker.read_version.compare(&checkv)) {
+    info = e.info;
+    if (pattrs)
+      *pattrs = e.attrs;
+    if (pmtime)
+      *pmtime = e.mtime;
+    return 0;
+  }
+
+  lderr(cct) << "Beware! The bucket info cache is inconsistent. This is an "
+             << "error that should be debugged. I am a nice machine, so I "
+             << "will try to recover." << dendl;
+
+  binfo_cache->invalidate(bucket_entry);
+  auto r = bucket_info_cache_miss(obj_ctx, info.bucket.tenant,
```

Comment:
Not crazy about having the whole get_bucket_logic() twice. This one is pretty much a superset of the other one, maybe consider some kind of a logic merge (while keeping the same interfaces)?